### ISSUE TITLE
DNS and openebs failures

### ISSUE THREAD

**Issue by harpermj66** [canonical/microk8s#3071]
Hello.

I have set up a three node cluster in a data centre hosted by ovhcloud and I have been having issues with coredns and probably related to that openebs pvc failures - restarts and ultimately failures. 

I have followed advice suggested here:

https://github.com/canonical/microk8s/issues/3043 

 ...and because I think I have multiple network interfaces here:
https://github.com/canonical/microk8s/issues/1511 
[report.tar.gz](https://github.com/canonical/microk8s/files/8499967/report.tar.gz)

I have also tried numerous times to make use of this article:

https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/

..And I have tried muliple versions and purged and started again many times. All left me with no improvement.

Any help would be much appreciated. I've attached my inpect report.

**Comment by harpermj66** [comment]
From another post I tried disabling HA to use flanneld and etcd. This seems to give me a stable system but I'm obviously missing high availability features.

**Comment by balchua** [comment]
In the inspect tarball, when i looked at the `cluster-info` from dqlite, it only contains 1 node, which has the ip ending with `197` node.

But looking at the pods deployed, i see there are 3 `calico-node` running, which indicates that kubernetes seems to see 3 nodes, but dqlite only has 1.

Can you also get the inspect tarball from the other 2 nodes?
Thanks

**Comment by harpermj66** [comment]
Thanks for getting back to me. 

I have since removed ha-cluster and so back to flannel and that sort of works.

I can purge microk8s and reinstall everything this weekend because I would like to find out what’s going on. Then I could send you the inspections from the other nodes if necessary.

What would a healthy cluster-info look like though?

**Comment by harpermj66** [comment]
I wiped the cluster an install ubuntu 20 to remove any OS static.
Also installed version 21 to avoid any late breaking issues.
I've created a 3 node cluster and enabled dns although it doesn't work of course.

I've attached all 3 nodes inspections. I hope there is a simple explanation :-)

Thanks again.
[node3154867.tar.gz](https://github.com/canonical/microk8s/files/8544010/node3154867.tar.gz)
[node3166183.tar.gz](https://github.com/canonical/microk8s/files/8544011/node3166183.tar.gz)
[node3173026.tar.gz](https://github.com/canonical/microk8s/files/8544012/node3173026.tar.gz)

**Comment by balchua** [comment]
Hi @harpermj66 

I checked the kubelite logs on the node `ns3154867` and i see these.  I dont know if these are the reason why DNS doesnt work.  But it looks like the apiserver couldn't reach out to the nodes via the node name.

```
Apr 22 20:42:59 ns3154867 microk8s.daemon-kubelite[20965]: E0422 20:42:59.439165   20965 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host"}: error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host
Apr 22 20:43:03 ns3154867 microk8s.daemon-kubelite[20965]: E0422 20:43:03.084944   20965 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host"}: error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host
Apr 22 20:43:07 ns3154867 microk8s.daemon-kubelite[20965]: E0422 20:43:07.336984   20965 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host"}: error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host
Apr 22 20:43:11 ns3154867 microk8s.daemon-kubelite[20965]: E0422 20:43:11.006572   20965 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host"}: error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host
Apr 22 20:43:13 ns3154867 microk8s.daemon-kubelite[20965]: E0422 20:43:13.697869   20965 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host"}: error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host
Apr 22 20:43:18 ns3154867 microk8s.daemon-kubelite[20965]: E0422 20:43:18.808962   20965 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host"}: error dialing backend: dial tcp: lookup ns3173026 on 213.186.33.99:53: no such host
```

Usually i add the node names on each server's `/etc/hosts`

**Comment by harpermj66** [comment]
I saw those as well and I did then set up the hostname ip address mappings in the /etc/hosts files on each node to try to get it to work.

The 213.186.33.99 ip is the nameserver for the data centre however and that only seems to map to fully qualified domain names not host names. For example ns3154867.ip-51-89-173.eu I could ping by default from another machine but not ns3154867. I think in my case the hostnames mappings isn't helping - not with calico at least.

I suppose I could try someting like ipaddr fqdn hostname in the etc/hosts file but I suspect that wont' work either.

I had a similar issue yesterday with my company data centre where I had only one node but I couldn't reach the internet from within a pod unless I disabled ha-cluster and reverted to flannel.

Another point to note is on a home network I never have these issues because the home network hostname resolution is automatic - so no messing with hosts files. 

This looks like it should be a very common issue.

**Comment by balchua** [comment]
I think you can map the fqdn to kubelet using the `hostname-override` in this file
`/var/snap/microk8s/current/args/kubelet`
See if you still see that error.

**Comment by harpermj66** [comment]
What I did.
sudo snap remove --purge microk8s

added --hostname-override=ns3154867.ip-51-89-173.eu to /var/snap/microk8s/current/args/kubelet for that node
added --hostname-override=ns3166183.ip-51-89-192.eu to /var/snap/microk8s/current/args/kubelet to the second node with that fqdn.

I then restarted the nodes and deleted the none fqdn hosts that existed for microk8s kubectl get nodes 

microk8s add-node in the first node
microk8s join from the second node as a worker.

The second node didn't  appear but I found that -hostname-override=ns3166183.ip-51-89-192.eu in the kubelet file for second node had been changed to --hostname-override=ns3154867.ip-51-89-173.eu 

I did this twice to confirm the behaviour.

**Comment by harpermj66** [comment]
This all seems like it should be rudimentary. I'll try to work it out for myself but I suspect that will take more time than I have available. 

Is there any consulting option I could consider to get paid help if I fail?

**Comment by balchua** [comment]
Im sure there is enterprise support.  @ktsakalozos will be able to direct you to the right place.

Im trying to understand the scenario, please bear with me.
1. You purged and reinstall microk8s on all 3 nodes.
2. On each node you added `--hostname-override`.
3. You join the nodes, then suddenly the `--hostname-override` becomes the node name you intend to join to?

**Comment by balchua** [comment]
Have another thing in mind.  Clean up microk8s (with purge).
Set the node hostname using `sudo hostnamectl set-hostname ns3166183.ip-51-89-192.eu`.
Then join them.  So we don't need to play with the `hostname-override` and let the system take the default of what the command `hostname` will give.

**Comment by harpermj66** [comment]
Yes, that's correct with regard to the --hostname-override. 
I actually already tried what you suggested with the hostname-ctl a couple of weeks ago and it didn't make dns resolution work, it just stripped the hostname at the first .

**Comment by balchua** [comment]
The dns resolution that doesn't work is within the cluster or it doesn't resolve outside?  Can you try the latest 1.23 version?  Like the edge?

**Comment by harpermj66** [comment]
It's within the cluster the dns doesn't work. Bear in mind I've had everything working by reverting to flannel --> removing ha-cluster

**Comment by harpermj66** [comment]
I will try the latest edge

**Comment by harpermj66** [comment]
I tried the latest edge version but it didn't fix the dns. From this site https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution I downloaded the pod for checking dns and ran:
kubectl exec -i -t dnsutils -- nslookup kubernetes.default

I get the following Error from server: error dialing backend: dial tcp: lookup ns3166183 on 213.186.33.99:53: no such host
This is a two node cluster and ns3166183 is the worker node that it the setup can't reach from the master.

213.186.33.99 is the domain name server for the data centre which does not store mappings for ns3166183. 

I guess the trick is still going to be to use the fqdn but as I said previously microk8s doen't seem to like those.
